src/test/spec/unified_runner/test_runner.rs

```diff
@@ -60,58 +64,74 @@ const SKIPPED_OPERATIONS: &[&str] = &[
 }
 
 impl TestRunner {
-    pub async fn new() -> Self {
+    pub(crate) async fn new() -> Self {
         Self {
             internal_client: TestClient::new().await,
-            entities: HashMap::new(),
-            fail_point_guards: Vec::new(),
+            entities: Default::default(),
+            fail_point_guards: Default::default(),
         }
     }
 
-    pub async fn new_with_connection_string(connection_string: &str) -> Self {
+    pub(crate) async fn new_with_connection_string(connection_string: &str) -> Self {
         #[cfg(all(not(feature = "sync"), not(feature = "tokio-sync")))]
         let options = ClientOptions::parse(connection_string).await.unwrap();
         #[cfg(any(feature = "sync", feature = "tokio-sync"))]
         let options = ClientOptions::parse(connection_string).unwrap();
         Self {
             internal_client: TestClient::with_options(Some(options)).await,
-            entities: HashMap::new(),
-            fail_point_guards: Vec::new(),
+            entities: Arc::new(RwLock::new(EntityMap::new())),
+            fail_point_guards: Arc::new(RwLock::new(Vec::new())),
         }
     }
 
-    pub async fn run_test(&mut self, test_file: TestFile, pred: impl Fn(&TestCase) -> bool) {
+    pub(crate) async fn run_test(
+        &self,
+        path: impl Into<Option<PathBuf>>,
+        test_file: TestFile,
+        pred: impl Fn(&TestCase) -> bool,
+    ) {
+        let test_runner = TestRunner::new().await;
+        let path = path.into();
+        let file_title = path
+            .as_ref()
+            .map(|p| p.display().to_string())
+            .unwrap_or_else(|| test_file.description.clone());
+
         if let Some(requirements) = test_file.run_on_requirements {
             let mut can_run_on = false;
             for requirement in requirements {
-                if requirement.can_run_on(&self.internal_client).await {
+                if requirement.can_run_on(&test_runner.internal_client).await {
                     can_run_on = true;
                 }
             }
             if !can_run_on {
-                log_uncaptured(format!(
-                    "Skipping file {}; client topology not compatible with test",
-                    test_file.description
+                file_level_log(format!(
+                    "Skipping file {}: client topology not compatible with test",
+                    file_title
                 ));
                 return;
             }
         }
 
-        log_uncaptured(format!("Running file {}", test_file.description));
+        log_uncaptured(format!(
+            "\n------------\nRunning tests from {}\n",
+            file_title
+        ));
 
         for test_case in test_file.tests {
             if let Some(skip_reason) = test_case.skip_reason {
                 log_uncaptured(format!(
-                    "Skipping test case {}: {}",
+                    "Skipping test case \"{}\": {}",
```

Comment:
Nit: when logging strings that need to be quoted, I tend to use `{:?}` to catch special characters as well.